### DISCUSSION TITLE
I want to help build features!

### DISCUSSION THREAD

**Discussion by hrodmn** [developmentseed/eoAPI#77] [General]
@vincentsarago I think eoAPI is the perfect foundation for some work that I am doing at NCX right now. I want to be able to a) build mosaic endpoints on the fly (I think you are already looking at this [here](https://github.com/developmentseed/eoAPI/pull/62), b) manipulate the asset/rescale/colormap parameters for a mosaic and have some more interactive features (layer control, basemap control, etc) while browsing the `map` endpoint (I think a lot of the code for asset display control already exists [here](https://github.com/developmentseed/eoAPI/blob/main/runtime/eoapi/raster/eoapi/raster/templates/stac-viewer.html)).

I would love to hear more about your vision for eoAPI and start contributing on actual features. Would you be interested in hopping on a call sometime so I can learn more about what you have in mind? Async is fine, too.

Thanks for all of your work on this and the whole family of cloud-native tools!

**Comment by vincentsarago** [comment]
Hi @hrodmn 🙏 

Yeah I have a lot of ideas about `frontend` for eoAPI (raster) services but lack of frontend knowledge/time/design skills 😄 

I started the `mosaic builder` because right now creating a STAC search cql query is still a pain AND I also wanted to ease the creation of the mosaic metadata (and are more usage of the metadata). 

That's said, that will not be an easy task but if you have ideas + time to work on this I'll be more than happy to help. I should make some sketch about some potential design and what would be needed/optional.

**Comment by hrodmn** [comment]
Sounds good, @vincentsarago! I don't have much frontend experience but I have some time to spend on this and would like to start learning more about interactive map design. If it's ok with you I'll start hacking on interactive features in the mosaic map endpoint. I will open up an issue where we can continue brainstorming.

**Reply by vincentsarago** [reply]
FYI the mosaic `/map` endpoint comes from https://github.com/stac-utils/titiler-pgstac/blob/main/titiler/pgstac/factory.py#L420-L491 and use the `index.html` template from `titiler.core` https://github.com/developmentseed/titiler/blob/main/src/titiler/core/titiler/core/templates/index.html 

Sadly you can't just override the templates as you can do in [`tipg`](https://github.com/developmentseed/eoAPI/blob/main/runtime/eoapi/vector/eoapi/vector/app.py#L36-L45) (I'll open an issue in titiler repo) so you'll need to do something like:

```python

templates = Jinja2Templates(  # type: ignore
    directory="",
    loader=jinja2.ChoiceLoader(
        [
            jinja2.PackageLoader("eoapi.raster", "templates"),
        ]
    ),
)

# https://github.com/developmentseed/titiler/blob/main/src/titiler/extensions/titiler/extensions/viewer.py#LL21C1-L42C1
@dataclass
class mosaicViewerExtension(FactoryExtension):
    """Add /map endpoint to the TilerFactory."""

    def register(self, factory: BaseTilerFactory):
        """Register endpoint to the tiler factory."""

        @factory.router.get("/{searchid}/map", response_class=HTMLResponse)
        @factory.router.get(
            "/{searchid}/{TileMatrixSetId}/map", response_class=HTMLResponse
        )
        def map_viewer(
            request: Request,
            searchid=Depends(factory.path_dependency),
            TileMatrixSetId: Literal["WebMercatorQuad"] = Query(
                "WebMercatorQuad",
                description="TileMatrixSet Name (default: 'WebMercatorQuad')",
            ),  # noqa
            tile_format: Optional[ImageType] = Query(
                None, description="Output image type. Default is auto."
            ),  # noqa
            tile_scale: int = Query(
                1, gt=0, lt=4, description="Tile size scale. 1=256x256, 2=512x512..."
            ),  # noqa
            minzoom: Optional[int] = Query(
                None, description="Overwrite default minzoom."
            ),  # noqa
            maxzoom: Optional[int] = Query(
                None, description="Overwrite default maxzoom."
            ),  # noqa
            layer_params=Depends(factory.layer_dependency),  # noqa
            dataset_params=Depends(factory.dataset_dependency),  # noqa
            pixel_selection: PixelSelectionMethod = Query(
                PixelSelectionMethod.first, description="Pixel selection method."
            ),  # noqa
            buffer: Optional[float] = Query(
                None,
                gt=0,
                alias="buffer",
                title="Tile buffer.",
                description="Buffer on each side of the given tile. It must be a multiple of `0.5`. Output **tilesize** will be expanded to `tilesize + 2 * buffer` (e.g 0.5 = 257x257, 1.0 = 258x258).",
            ),  # noqa
            post_process=Depends(factory.process_dependency),  # noqa
            rescale: Optional[List[Tuple[float, ...]]] = Depends(
                RescalingParams
            ),  # noqa
            color_formula: Optional[str] = Query(
                None,
                title="Color Formula",
                description="rio-color formula (info: https://github.com/mapbox/rio-color)",
            ),  # noqa
            colormap=Depends(factory.colormap_dependency),  # noqa
            render_params=Depends(factory.render_dependency),  # noqa
            pgstac_params: PgSTACParams = Depends(),  # noqa
            backend_params=Depends(factory.backend_dependency),  # noqa
            reader_params=Depends(factory.reader_dependency),  # noqa
            env=Depends(factory.environment_dependency),  # noqa
        ):
            """Return a simple map viewer."""
            tilejson_url = factory.url_for(
                request, "tilejson", searchid=searchid, TileMatrixSetId=TileMatrixSetId
            )
            if request.query_params._list:
                tilejson_url += f"?{urlencode(request.query_params._list)}"

            tms = factory.supported_tms.get(TileMatrixSetId)
            return templates.TemplateResponse(
                name="index.html",
                context={
                    "request": request,
                    "tilejson_endpoint": tilejson_url,
                    "tms": tms,
                    "resolutions": [tms._resolution(matrix) for matrix in tms],
                },
                media_type="text/html",
            )


###############################################################################
# MOSAIC Endpoints
# https://github.com/developmentseed/eoAPI/blob/main/runtime/eoapi/raster/eoapi/raster/app.py
mosaic = MosaicTilerFactory(
    optional_headers=optional_headers,
    router_prefix="/mosaic",
    add_statistics=True,
    add_map_viewer=False,
    add_mosaic_list=True,
    extensions=[
        mosaicViewerExtension(),
    ],
)
app.include_router(mosaic.router, tags=["Mosaic"], prefix="/mosaic")
```

But I'll try to make a new release of titiler and titiler.pgstac to allow full customization of the `templates` variable to user can easily override the html code!
